sycl/source/detail/platform_impl.cpp

```diff
@@ -140,14 +141,29 @@ std::vector<platform> platform_impl::get_platforms() {
 // This function matches devices in the order of backend, device_type, and
 // device_num.
 static void filterDeviceFilter(std::vector<RT::PiDevice> &PiDevices,
-                               const plugin &Plugin) {
+                               RT::PiPlatform Platform) {
   device_filter_list *FilterList = SYCLConfig<SYCL_DEVICE_FILTER>::get();
   if (!FilterList)
     return;
+  const std::lock_guard<std::mutex> Guard(
+      GlobalHandler::instance().getPluginsMutex());
+  std::vector<plugin> &Plugins = GlobalHandler::instance().getPlugins();
+  if (Plugins.size() == 0)
+    RT::initialize();
```

Comment:
This function locks `GlobalHandler::instance().getPluginsMutex()` and calls `RT::initialize` which also locks `GlobalHandler::instance().getPluginsMutex()`. This should lead to deadlock.

Probably we could move call to `RT::initialize` to `GlobalHandler::instance().getPlugins()` function.

Reply:
Thanks for pointing out this deadlock.
I moved RT::initialize into getPlugins() as you suggested.

Reply:
Removing lock returns us back to the previous problem - races between:
1. RT::initialize(); - accesses the plugins vector for writing
2. if (Plugins.empty()) - accesses the plugins vector for reading
3. find_if - accesses the plugins vector for reading
4. containsPiPlatform - accesses plugin objects for reading
5. getStartingDeviceId - accesses plugins objects for reading and writing
6. setLastDeviceId - accesses plugin objects for reading and writing

As far as I can see all this functions can be called from different threads at the same time which is not thread safe.

Reply:
Thanks again.
By the way, I did not need to call `getPlugins` separately to get the plugins vector.
RT::initialize() already uses a lock to block all reading threads from reading plugins vector until the plugins vector is created and return the vector to the caller.
So, I simply can use the return vector without calling `getPlugins` again.

Reply:
It seems there are still races between:
int DeviceNum = Plugin.getStartingDeviceId(Platform);
Plugin.setLastDeviceId(Platform, DeviceNum);
Plugins[i].resetLastDeviceIds();
Plugins[i].getPlatformId(PiPlatform);

Reply:
Done. Thanks.